File: packages/sdk/java/core/src/main/java/dev/bugwatch/StackFrame.java
package dev.bugwatch;

import com.google.gson.annotations.SerializedName;

/**
 * Represents a single frame in a stack trace.
 */
public record StackFrame(
        String filename,
        @SerializedName("function") String function,
        @SerializedName("lineno") int lineno,
        @SerializedName("colno") Integer colno,
        @SerializedName("abs_path") String absPath,
        @SerializedName("in_app") boolean inApp,
        String module
) {

    /**
     * Builder for constructing StackFrame instances.
     */
    public static final class Builder {
        private String filename;
        private String function;
        private int lineno;
        private Integer colno;
        private String absPath;
        private boolean inApp;
        private String module;

        public Builder filename(String filename) {
            this.filename = filename;
            return this;
        }

        public Builder function(String function) {
            this.function = function;
            return this;
        }

        public Builder lineno(int lineno) {
            this.lineno = lineno;
            return this;
        }

        public Builder colno(Integer colno) {
            this.colno = colno;
            return this;
        }

        public Builder absPath(String absPath) {
            this.absPath = absPath;
            return this;
        }

        public Builder inApp(boolean inApp) {
            this.inApp = inApp;
            return this;
        }

        public Builder module(String module) {
            this.module = module;
            return this;
        }

        public StackFrame build() {
            return new StackFrame(filename, function, lineno, colno, absPath, inApp, module);
        }
    }
}

File: packages/sdk/php/src/Types/StackFrame.php
<?php

declare(strict_types=1);

namespace Bugwatch\Types;

final class StackFrame
{
    public function __construct(
        public readonly ?string $filename = null,
        public readonly ?string $function = null,
        public readonly ?int $lineno = null,
        public readonly ?int $colno = null,
        public readonly ?string $absPath = null,
        public readonly bool $inApp = true,
        public readonly ?string $contextLine = null,
    ) {}

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return array_filter([
            'filename' => $this->filename,
            'function' => $this->function,
            'lineno' => $this->lineno,
            'colno' => $this->colno,
            'abs_path' => $this->absPath,
            'in_app' => $this->inApp,
            'context_line' => $this->contextLine,
        ], fn (mixed $value): bool => $value !== null);
    }
}
